refactor: new auth provider

mateo-ivc opened the pull request:
<!--
Please make sure the title of your PR starts with a semantic prefix:

build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
ci: Changes to CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
chore: Changes which doesn't change source code or tests e.g. changes to the build process, auxiliary tools, libraries
docs: Documentation only changes
feat: A new feature
fix: A bug fix
perf: A code change that improves performance
refactor: A code change that neither fixes a bug nor adds a feature
revert: Revert something
style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
test: Adding missing tests or correcting existing tests
-->

## Description
<!-- Explain the changes you’ve made. It doesn’t need to be fancy and you don’t have to get too technical. -->

## Changelog
<!-- Please provide a detailed list of the changes you have made to the codebase. -->

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] The light- and dark-theme are both supported and tested
- [ ] The design was implemented and is responsive for all devices and screen sizes
- [ ] The application was tested in the most commonly used browsers (e.g. Chrome, Firefox, Safari)

## (Optional) Visual Changes
<!-- If available, please provide a before and after screenshot of your UI related changes. -->

Commit:
new auth provider

# Conflicts:
#	server/src/api/login.go
#	server/src/api/router.go
#	server/src/go.mod
#	server/src/go.sum

Commit:
test

Lennart01 committed:
fix: get a valid tls cert

Commit:
debug

Commit:
linter

Commit:
auth opts changes

Commit:
remove issuer

Commit:
logger for auth

Commit:
disable secure cookie

Commit:
authHandler when establishing connection

Commit:
fix routing

Commit:
fix routing

Commit:
fix routing

Commit:
fix linter

Commit:
fix linter

mateo-ivc committed:
redirect not working

mateo-ivc committed:
reset login url

mateo-ivc committed:
working google prov

mateo-ivc committed:
debug microsoft oauth

mateo-ivc committed:
linter things

mateo-ivc committed:
oicd stuff and user extraction

mateo-ivc committed:
refactor auth

mateo-ivc committed:
api tests

mateo-ivc committed:
tests

mateo-ivc committed:
tests

mateo-ivc committed:
fix open settings after login

mateo-ivc committed:
golang linter

mateo-ivc committed:
check signiture of token sent from provider

mateo-ivc committed:
fix tests

mateo-ivc committed:
change tracer to lower case

mateo-ivc committed:
Doku and adjustment to .env

mateo-ivc committed:
Doku and adjustment to .env

mateo-ivc committed:
add comments

Planlos5000 commented on server/src/api/column_templates.go:
Revert the file. It should still be var tracer trace.Tracer = otel.Tracer("scrumlr.io/server/api")

Planlos5000 commented on server/src/api/feedback.go:
Revert the file. It should still be var tracer trace.Tracer = otel.Tracer("scrumlr.io/server/api")

Planlos5000 commented on server/src/auth/service.go:
```diff
@@ -0,0 +1,437 @@
+  return a.oauthConfigs[provider]
+}
+
+func (a *AuthConfiguration) initializeProviders() error {
```
Add the otel tracer and maybe add some debug log

mateo-ivc committed:
add apple

mateo-ivc committed:
fix merge conflicts

mateo-ivc committed:
remove unused otel comments

mateo-ivc committed:
resolve comments

wischoepke commented on server/src/auth/api.go:
```diff
@@ -0,0 +1,159 @@
+		return
+	}
+	nonce := base64.URLEncoding.EncodeToString(nonceBytes)
+	returnURL := r.URL.Query().Get("state")
```
Possible Open Redirect attack: You need to validate returnURL before using it. Ensure it only contains a relative path (e.g., starts with /) or matches a strict whitelist of allowed domains.

mateo-ivc committed:
resolve issues

mateo-ivc committed:
move objects into own package

mateo-ivc committed:
validate return url

mateo-ivc committed:
fix tests

mateo-ivc committed:
cleanup dependecies

mateo-ivc committed:
move http package to auth package

mateo-ivc committed:
add helper for cookies

mateo-ivc committed:
rebase & change cookie to strict mode

mateo-ivc committed:
fix cookie, provider issue

mateo-ivc committed:
linter stuff

mateo-ivc committed:
remove duplicate

mateo-ivc committed:
minor fixes

malterusko-ino committed:
chore: go mod tidy

malterusko-ino committed:
refactor: replaced new-functions to the top

malterusko-ino committed:
refactor: remove redundant location header in callback

The manually set Location header was being overwritten by the subsequent http.Redirect call. Since http.Redirect handles header setting internally, the manual construction was unnecessary.

malterusko-ino committed:
refactor: simplify oauth state handling and redirect logic

malterusko-ino committed:
refactor: removed unused test suite

malterusko-ino committed:
fix: added missing return

malterusko-ino committed:
fix: ensure single response in oauth flow by removing redundant throw

malterusko-ino committed:
feat: implemented safe redirect validation and host whitelist

malterusko-ino committed:
feat: wire redirect host whitelist to CLI and initializer

malterusko-ino committed:
docs: document allowed redirect hosts configuration

malterusko-ino committed:
chore: go mod tidy

malterusko-ino committed:
refactor: use hostnames for redirect allowlist validation

malterusko-ino committed:
Merge remote-tracking branch 'origin/main' into refactor/mi/auth

malterusko-ino committed:
chore: go mod tidy

malterusko-ino committed:
docs: fixed naming

malterusko-ino committed:
chore: configured auth callback and redirect hostnames for local development

malterusko-ino committed:
chore: added session-secret to config_example and Makefile

malterusko-ino committed:
Merge remote-tracking branch 'origin/main' into refactor/mi/auth

malterusko-ino committed:
chore: go mod tidy

malterusko-ino commented on server/src/auth/api.go:
```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		span.SetStatus(codes.Error, "failed to handle callback")
+		span.RecordError(err)
+		log.Errorw("failed to handle callback", "err", err)
```
Within `common.Throw`, HTTP headers are set by the `go-chi/render` library.

malterusko-ino commented on server/src/auth/api.go:
```diff
@@ -0,0 +1,257 @@
+
+	targetURL := parts[1]
+	if target, safe := api.isSafeRedirect(targetURL); safe {
+		http.Redirect(w, r, target, http.StatusSeeOther)
```
The issue was resolved with the introduction of the isSafeRedirect validation logic. This function ensures that redirect targets are either:

1. Safe relative paths (starting with `/`, but excluding `//` to prevent protocol-relative redirects).
2. Are permitted external hosts that are matched against an explicit allowlist (allowRedirectHostnames).

Additionally, backslashes are normalized, and evasion attempts (e.g., leading backslashes) are intercepted.

malterusko-ino committed:
chore: revert k8s/deployment.yaml

ffischbach commented on server/src/auth/service.go:
```diff
@@ -0,0 +1,414 @@
+
+  tokenString, _ := a.Sign(map[string]interface{}{"id": internalUser.ID})
+
+  cookie := CreateCookie("jwt", tokenString, "/", 121*24*60*60, nil) //maxAge = 21 Days
```
`121*24*60*60` = 121 Days not 21...

ffischbach commented on server/src/main.go:
```diff
@@ -435,31 +453,44 @@ func run(ctx *cli.Context) error {
+	authService := initializer.InitializeAuthService(providersMap, unsafeKeyWithNewlines, keyWithNewlines, userService)
 	if err != nil {
 		return fmt.Errorf("unable to setup authentication: %w", err)
 	}
```
Does this err check work as expected? initializer.InitializeAuthService returns nil instead of an error - like before.

malterusko-ino replied:
For now, I've set the error check to nil. I'll add this to the error-handling ticket and implement an appropriate error.

ffischbach commented on server/src/auth/api.go:
```diff
@@ -0,0 +1,250 @@
+	}
+
+	//format: nonce__returnURL
+	parts := strings.Split(state, "__")
```
Since we expect the state to be split in (at least) three parts, we could add a check like ```  if len(parts) < 2 {common.Throw(w, r, common.BadRequestError(errors.New("invalid state format"))) return} ``` to prevent a index out of bounce exception. Event if it should never happen...

malterusko-ino replied:
Good catch, thanks!

ffischbach commented on server/src/auth/service.go:
```diff
@@ -0,0 +1,414 @@
+    internalUser, err = a.userService.CreateGitHubUser(ctx, userInfo.Ident, userInfo.Name, userInfo.AvatarURL)
+  case common.TypeOIDC:
+    internalUser, err = a.userService.CreateOIDCUser(ctx, userInfo.Ident, userInfo.Name, userInfo.AvatarURL)
+  case common.Microsoft:
```
Is `AZURE_AD` missing here? 
```
const (
	// Anonymous users don't require a registration
	Anonymous AccountType = "ANONYMOUS"

	// Google users registered on Google
	Google AccountType = "GOOGLE"

	// Microsoft users registered on Microsoft
	Microsoft AccountType = "MICROSOFT"

	// AzureAd users registered on Azure AD
	AzureAd AccountType = "AZURE_AD"

	// GitHub users registered on GitHub
	GitHub AccountType = "GITHUB"

	// Apple users registered on Apple
	Apple AccountType = "APPLE"

	// TypeOIDC users registered on OIDC
	TypeOIDC AccountType = "OIDC"
)
```

ffischbach replied:
```
case common.AzureAd:
   internalUser, err = a.userService.CreateAzureAdUser(ctx, userInfo.Ident, userInfo.Name, userInfo.AvatarURL)
```

ffischbach commented on server/src/auth/service.go:
```diff
@@ -0,0 +1,414 @@
+    return nil, common.InternalServerError
+  }
+
+  tokenString, _ := a.Sign(map[string]interface{}{"id": internalUser.ID})
```
shouldn't we handle a potential signing err here, to avoid returning 200 with an empty token?

ffischbach commented on server/src/auth/router.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+func (r *Router) RegisterRoutes() chi.Router {
+	router := chi.NewRouter()
+
+	router.Group(func(router chi.Router) {
+		router.Delete("/", r.authAPI.Logout)
+		router.Post("/anonymous", r.authAPI.SignInAnonymously)
+
+		router.Route("/{provider}", func(router chi.Router) {
+			router.Get("/", r.authAPI.BeginAuth)
+			router.Get("/callback", r.authAPI.Callback)
+		})
+	})
+
+	return router
+}
```
Are we aware, that we are removing the `AnonymousLoginDisabledContext` middleware?

<img width="988" height="243" alt="Image" src="https://github.com/user-attachments/assets/671bcb5c-c81c-4931-8569-011e5030aa8b" />

malterusko-ino replied:
Good catch! I've added the middleware back.

malterusko-ino committed:
fix: calculation of cookie max age

malterusko-ino committed:
fix: handle potential signing errors during JWT creation

malterusko-ino committed:
fix: correct Azure AD claim mapping and improve robustness

malterusko-ino committed:
fix: prevent index out of bounds by validating oauth state

malterusko-ino committed:
fix: restore anonymous login protection middleware

malterusko-ino committed:
fix: check for nil on auth service initialization

malterusko-ino committed:
fix: removed unused var
